### PR TITLE
Add Application Interaction Lecture 

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

File: content/chapters/app-interact/lecture/slides.mdpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+## OS - The Lawmaker
+
+* The OS offers the primitives for app-app intercation
+* The OS is like a lawmaker the dictates how the interaction takes place
+* TODO: image with user-app, app-app interface on top of OS (same/remote system)
+
+---
+
+## App Classification
```
Comment (teodutu) [inline content/chapters/app-interact/lecture/slides.mdpp:144]: You could build upon and expand this image: https://github.com/open-education-hub/operating-systems/blob/master/content/chapters/compute/lecture/media/app-process-thread.svg

File: content/chapters/app-interact/lecture/slides.mdpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+----
+
+## Shared memory
+
+- TODO: img
+- apps have a common memory region they use to communicate
+- OS sets up the memory
+- requires synchronization
+- used primarily by intra-app applications
```
Comment (teodutu) [inline content/chapters/app-interact/lecture/slides.mdpp:181]:
```suggestion
- used primarily by intra-app communication
```

File: content/chapters/app-interact/lecture/slides.mdpp [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+- OS does the transmitting
+- requires communication protocol
+- used for inter-application interaction, but could be used for intra-app communication also
+
+----
+
+## Shared memory
+
+- TODO: img
+- apps have a common memory region they use to communicate
```
Comment (teodutu) [inline content/chapters/app-interact/lecture/slides.mdpp:178]: Be more specific that different processes map the same frames. You can use [this demo](https://github.com/open-education-hub/operating-systems/blob/master/content/chapters/compute/lab/content/arena.md#shared-memory) as a starting point.

**Review by teodutu — CHANGES_REQUESTED** [review]
I made a first pass through the slides. The skeleton is good, but be consistent with lists: either use `*` or `-`. I'd say `-`.

File: content/chapters/app-interact/lecture/slides.mdpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+----
+
+### Exhibit 4 - WhatsApp application
+
+* There is interaction between the main client application and other applications: photos, location, youtube etc.
+* There is also interaction with the web server when sending messages.
+
+----
+
+### What does it mean that applications interact?
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply content/chapters/app-interact/lecture/slides.mdpp:113]: The person that delivers the content may touch on this at the beginning, but then it should re-explained after seeing some examples.

File: content/chapters/app-interact/lecture/demo/comm-channels/receive_fifo.c [hunk 5]
```diff
@@ -0,0 +1,41 @@
+#include <unistd.h>
```
Comment (teodutu) [inline content/chapters/app-interact/lecture/demo/comm-channels/receive_fifo.c:1]:
```suggestion
/* SPDX-License-Identifier: BSD-3-Clause */

#include <unistd.h>
```

File: content/chapters/app-interact/lecture/slides/classification.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+- IPC mechanisms (pipes, message queues, sockets) to interface between processes
+
+- Postfix
+- GitLab
+
+----
+
+### Postfix
+
+![Postfix](./media/postfix-architecture.svg)
```
Comment (teodutu) [inline content/chapters/app-interact/lecture/slides/classification.md:84]: The writing in this image is very small. Make it larger. The background is also transparent. This sucks if you render it against a dark background. Add a white square behind the drawing to force the background to be white.

**Review by teodutu — CHANGES_REQUESTED** [review]
I like your slides so far, especially the images you've added. Consider changing all paths/ports hardcoded into your demos to command-line arguments for flexibility and increased interaction.

**Comment by teodutu** [comment]
In addition, be mindful of the linter errors. They seem relevant and rather easy to fix.

File: content/chapters/app-interact/lecture/demo/comm-channels/receive_fifo.c [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+#include <fcntl.h>
+#include <sys/stat.h>
+
+#include "utils/utils.h"
+
+static const char fifo_path[] = "fifo_channel";
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/chapters/app-interact/lecture/demo/comm-channels/receive_fifo.c:9]: I would leave this as a TODO item for student contributions.

File: content/chapters/app-interact/lecture/demo/comm-channels/receive_net_dgram_socket.c [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+#include <sys/socket.h>
+#include <netinet/in.h>
+
+#include "utils/utils.h"
+
+#ifndef BUFSIZ
+#define BUFSIZ		256
+#endif
+
+unsigned short port = 4242;
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/chapters/app-interact/lecture/demo/comm-channels/receive_net_dgram_socket.c:16]: I would leave this as a TODO item for student contributions.

File: content/chapters/app-interact/lecture/slides/classification.md (hunk 6 above)
Comment (razvand) [reply content/chapters/app-interact/lecture/slides/classification.md]: I made it white. The text is not that relevant, I wouldn't bother about it.

File: content/chapters/app-interact/lecture/slides/classification.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+### App Classification
+
+- Single-process single-threaded
+- Multi-threaded
+- Multi-process
+- Multi-system
+- These apps use OS primitives to communicate between components (threads, processes)
+
+----
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/chapters/app-interact/lecture/slides/classification.md:29]: Doesn't make sense to move them elsewhere. Images are tied to a particular app class. I will leave them like this.

File: content/chapters/app-interact/lecture/slides/os-roles.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+- Address / Identity of processes
+  - Process ID
+  - Hostname, port
+- Communication channel
+  - Buffer to store data
+  - "Endpoints" for processes to send / receive data from buffer
+- Address / Identity of channel
+  - Path in filesystem
+  - Hostname, port, path
+  - ID
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/chapters/app-interact/lecture/slides/os-roles.md:18]: Did

**Review by teodutu — APPROVED** [review]
I'm OK with what you have so far. See if you find a quick solution so you no longer need to cast unused arguments to `void`.

File: content/chapters/app-interact/lecture/slides/channels.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+### Synchronization
+
+![Synchronization](./media/synchronization-generated.gif)
+
+----
+
+### Interruption
+
+![Interruption](./media/interruption-generated.gif)
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/chapters/app-interact/lecture/slides/channels.md:52]: @teodutu, fixed using `convert`.

File: content/chapters/app-interact/lecture/demo/comm-channels/CPPLINT.cfg [hunk 12]
```diff
@@ -0,0 +1 @@
+filter=-runtime/printf
```
Comment (teodutu) [inline content/chapters/app-interact/lecture/demo/comm-channels/CPPLINT.cfg:1]: This filter is disabled globally. No need to disable it again here.

**Review by teodutu — CHANGES_REQUESTED** [review]
In addition to my inline comments, be consistent with the `SPDX` comments. You wrote some of them like this:

```c
// SPDX-License-Identifier: BSD-3-Clause
```

while others look like this:

```c
/*
 * SPDX-License-Identifier: BSD-3-Clause
 */
```

Why not just use regular single-line C-style comments (`/* ... */`) instead?

**Review by teodutu — APPROVED** [review]
Changed my review to "Approve" just in case you need to merge this quickly.

File: content/chapters/app-interact/lecture/demo/fibonacci-server/CPPLINT.cfg [hunk 13]
```diff
@@ -0,0 +1 @@
+filter=-build/include
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/chapters/app-interact/lecture/demo/fibonacci-server/CPPLINT.cfg:1]: It complains that the header is not included in some sort of full path:

```console
razvan@yggdrasil:~/.../lecture/demo/fibonacci-server$ cpplint *.{c,h}
connection.c:5:  content/chapters/app-interact/lecture/demo/fibonacci-server/connection.c should include its header file content/chapters/app-interact/lecture/demo/fibonacci-server/connection.h. Relative paths like . and .. are not allowed.  [build/include] [5]
```

If using `#include "connection.h"` you get:

```console
razvan@yggdrasil:~/.../lecture/demo/fibonacci-server$ cpplint *.{c,h}
connection.c:21:  Include the directory when naming header files  [build/include_subdir] [4]
```

Mother-f***ng `cpplint`.

File: content/chapters/app-interact/lecture/demo/fibonacci-server/connection.c [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+#include "utils/log/log.h"
+#include "utils/sock/sock_util.h"
+#include "utils/utils.h"
+
+#include "./connection.h"
+
+/*
+ * Compute Fibonacci number.
+ */
+
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/chapters/app-interact/lecture/demo/fibonacci-server/connection.c:26]: I typically leave an empty line after a function comment. We can talk about coding style. I would leave it like this, for consistency.

[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/app-interact/lecture/demo/fibonacci-server/connection.c:24]: Ok. I don't really give a shit about this, but the rest of the code doesn't leave any empty lines.

File: content/chapters/app-interact/lecture/demo/fibonacci-server/CPPLINT.cfg (hunk 13 above)
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/app-interact/lecture/demo/fibonacci-server/CPPLINT.cfg:1]: :upside_down_face: I see.

**Review by teodutu — APPROVED** [review]
I declare myself satisfied.

File: content/chapters/app-interact/lecture/demo/fibonacci-server/connection.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (razvand) [reply content/chapters/app-interact/lecture/demo/fibonacci-server/connection.c:24]: When I use function comments, I leave an empty line after the comment and before the actual function. What is the place where this isn't the case?

[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/app-interact/lecture/demo/fibonacci-server/connection.c:24]: Pretty much any code not written by you.

[the remark this answers is not in the document]
Comment (razvand) [reply content/chapters/app-interact/lecture/demo/fibonacci-server/connection.c:24]: We'll have a discussion on comment placing.

**Pull request merged**